daffodil-runtime1-unparser/src/main/scala/edu/illinois/ncsa/daffodil/processors/unparsers/IBM4690PackedDecimalUnparsers.scala

```diff
@@ -0,0 +1,136 @@
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the names of Tresys Technology, nor the names of its contributors
+ *     may be used to endorse or promote products derived from this Software
+ *     without specific prior written permission.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.processors.unparsers
+
+import edu.illinois.ncsa.daffodil.processors.parsers.HasKnownLengthInBits
+import edu.illinois.ncsa.daffodil.processors.ElementRuntimeData
+import edu.illinois.ncsa.daffodil.processors.ParseOrUnparseState
+import edu.illinois.ncsa.daffodil.util.Maybe._
+import edu.illinois.ncsa.daffodil.util.DecimalUtils._
+import java.lang.{ Number => JNumber, Long => JLong }
+import java.math.{ BigDecimal => JBigDecimal, BigInteger => JBigInteger }
+import edu.illinois.ncsa.daffodil.exceptions.Assert
+import edu.illinois.ncsa.daffodil.processors.parsers.HasRuntimeExplicitLength
+import edu.illinois.ncsa.daffodil.processors.Evaluatable
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.LengthUnits
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.YesNo
+import edu.illinois.ncsa.daffodil.io.DataOutputStream
+
+abstract class IBM4690PackedNumberBaseUnparser(e: ElementRuntimeData)
+  extends PrimUnparserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  protected def putNumber(dos: DataOutputStream, number: JNumber, nBits: Int): Boolean
+
+  def unparse(state: UState): Unit = {
+    val nBits = getBitLength(state)
+    val node = state.currentInfosetNode.asSimple
+    val value = node.dataValue.asInstanceOf[JNumber]
+    val dos = state.dataOutputStream
+
+    val res =
+      if (nBits > 0) {
+        putNumber(dos, value, nBits)
+      } else {
+        true
+      }
+
+    if (!res) {
+      Assert.invariant(dos.maybeRelBitLimit0b.isDefined)
+      UnparseError(One(state.schemaFileLocation), One(state.currentLocation), "Insufficient space to unparse element %s, required %s bits, but only %s were available.",
+        e.dpathElementCompileInfo.namedQName.toPrettyString, nBits, dos.maybeRelBitLimit0b.get)
+    }
+  }
+
+}
+
+abstract class IBM4690PackedIntegerBaseUnparser(e: ElementRuntimeData, signed: Boolean)
+  extends IBM4690PackedNumberBaseUnparser(e) {
+
+  override def putNumber(dos: DataOutputStream, value: JNumber, nBits: Int): Boolean = {
+    dos.putByteArray(IBM4690FromBigInteger(new JBigInteger(value.toString), nBits), nBits)
```

Comment:
Same question here. Can we avoid the JBigInteger by having IBM4690FromBigInteger accept a String?